### PR TITLE
refactor: rework hook registration

### PR DESCRIPTION
**Pull request by ProdPreva1l**
New way to register a hook:
```java
public class MyFancyHook extends Hook {
    @PluginHook(id = "MyFancyHook", register = PluginHook.Register.DELAYED, platform = "bukkit")
    public MyFancyHook(@NotNull HuskTowns plugin) {
        super(plugin);
    }
}
```

### CODE REVIEW

File: common/src/main/java/net/william278/husktowns/hook/HookManager.java [hunk 1]
```diff
@@ -0,0 +1,56 @@
...
+    protected final static HashSet<Hook> registeredHooks = new HashSet<>();
+
+    protected HookManager(@NotNull HuskTowns plugin) {
+        this.plugin = plugin;
+    }
+
+    public abstract void registerOnLoad();
+
+    public abstract void registerOnEnable();
+    public abstract void registerDelayed();
```
Comment (WiIIiam278) [inline common/src/main/java/net/william278/husktowns/hook/HookManager.java:40]: nit: missing whitespace (newline)?

File: common/src/main/java/net/william278/husktowns/hook/HookManager.java [hunk 2]
```diff
@@ -0,0 +1,56 @@
...
+import java.util.Objects;
+import java.util.Optional;
+
+public abstract class HookManager {
+    protected final HuskTowns plugin;
+    protected final static HashSet<Hook> registeredHooks = new HashSet<>();
+
+    protected HookManager(@NotNull HuskTowns plugin) {
+        this.plugin = plugin;
+    }
```
Comment (WiIIiam278) [inline common/src/main/java/net/william278/husktowns/hook/HookManager.java:35]: lombok this: annotate the class with @AllArgsConstructor

File: bukkit/src/main/java/net/william278/husktowns/hook/BukkitHookManager.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+                .filter((hook) -> hook.getHookInfo().register() == PluginHook.Register.ON_ENABLE)
+                .forEach((hook) -> {
+                    hook.enable();
+                    loaded.getAndIncrement();
+                });
+        plugin.log(Level.INFO, "Successfully loaded %s hooks".formatted(loaded.get()));
+    }
+
+    @Override
+    public void registerDelayed() {
```
Comment (WiIIiam278) [inline bukkit/src/main/java/net/william278/husktowns/hook/BukkitHookManager.java:60]: I struggle (and shudder) to think of a case where we'd use this but seems reasonable to include.

Comment (ProdPreva1l) [reply bukkit/src/main/java/net/william278/husktowns/hook/BukkitHookManager.java:60]: yeah i pray we dont need it but ive had to do it before with things like anticheat hooks

File: common/src/test/java/net/william278/husktowns/hook/maphooks/PlanHookTests.java
[the remark this answers is not in the document]
Comment (ProdPreva1l) [reply common/src/test/java/net/william278/husktowns/hook/maphooks/PlanHookTests.java:1]: woops lol

**Comment by WiIIiam278** [comment]
Looks like this causes compile to fail during testing? Reflection might be breaking something here.

**Comment by ProdPreva1l** [comment]
i think it has to with worldguardhook, that was the issue last time

**Comment by WiIIiam278** [comment]
it might be the Mockito trying to load the WG hook and going nuts for some reason.
is it crashing trying to load WG classes?

**Comment by ProdPreva1l** [comment]
> it might be the Mockito trying to load the WG hook and going nuts for some reason.
> 
> is it crashing trying to load WG classes?

i havent the slightest clue as to whats going wrong, maybe some bad accessors?

**Comment by WiIIiam278** [comment]
I did find that you're registering each hook twice

**Comment by ProdPreva1l** [comment]
> I did find that you're registering each hook twice

WHUUTT, lemme check this

**Comment by ProdPreva1l** [comment]
Bump

**Comment by ProdPreva1l** [comment]
Is there reviews that i cant see or am i tweaking bc i see "1 change requested" but no actual comments

**Review by WiIIiam278 — APPROVED** [review]

**Pull request merged**